File: 105.construct-binary-tree-from-preorder-and-inorder-traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    TreeNode* buildTree(vector<int>& preorder, vector<int>& inorder, int pre_left, int in_left, int len) {
        if (len == 0) {
            return NULL;
        }
        int val = preorder[pre_left];
        TreeNode* root = new TreeNode(val);
        int root_pos;
        for (int i = in_left; i < in_left + len; i++) {
            if (inorder[i] == val) {
                root_pos = i;
                break;
            }
        }
        int left_len = root_pos - in_left;
        root->left = buildTree(preorder, inorder, pre_left + 1, in_left, left_len);
        root->right = buildTree(preorder, inorder, pre_left + 1 + left_len, root_pos + 1, len - left_len - 1);
        return root;
    }

    TreeNode* buildTree(vector<int>& preorder, vector<int>& inorder) {
        return buildTree(preorder, inorder, 0, 0, preorder.size());
    }
};

File: 105.construct-binary-tree-from-preorder-and-inorder-traversal.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {number[]} preorder
 * @param {number[]} inorder
 * @return {TreeNode}
 */
var buildTree = function(preorder, inorder) {
    if (preorder.length == 0) {
        return null;
    }
    let root = new TreeNode(preorder[0]);
    let pos = inorder.indexOf(preorder[0]);
    root.left = buildTree(preorder.slice(1, pos + 1), inorder.slice(0, pos));
    root.right = buildTree(preorder.slice(pos + 1), inorder.slice(pos + 1));
    return root;
};